datapath = "/mnt/project/Data/qc_geno/"

library(snpStats)
library(SNPRelate)
library(plyr)

start.time<-Sys.time()
for(chr in 1:22){
  print(paste0("start: chr",chr))
  ## read in PLINK files to create a list
  # the .bed .bim .fam files are the output of PLINK
  geno <- read.plink(bed = paste0(datapath, 'ukb_qc_chr',chr,'.bed'), 
                     bim = paste0(datapath, 'ukb_qc_chr',chr,'.bim'),
                     fam = paste0(datapath, 'ukb_qc_chr',chr,'.fam'), na.strings = c("-9"))
  
  ## obtain genotype SnpMatrix from generated list
  genotype <- geno$genotypes
  
  ## obtain SNP information table from the list
  genoBIM <- geno$map
  colnames(genoBIM) <- c("chr", "SNP", "gen.dist", "position", "A1", "A2")
  
  ## obtain Samples information table 
  genoFAM <- geno$fam
  rm(geno)
  
  ## only keep SNPs on chr1
  genotype_chr = genotype[,genoBIM$SNP[genoBIM$chr == chr]]
  rm(genotype)
  
  # -------------------------- QCs for SNPs -------------------------
  
  ## SNP level filtering
  snpsum.col <- col.summary(genotype_chr)
  
  ## set threshold
  call <- 0.9
  minor <- 0.01
  use <- with(snpsum.col, (!is.na(MAF) & MAF > minor) & (Call.rate > call))
  use[is.na(use)] <- FALSE # remove NAs
  
  cat(ncol(genotype_chr) - sum(use), "SNPs will be removed due to low MAF or call rate. \n")
  
  ## subset SNPs that passed the criteria of MAF and call rate
  genotype_chr <- genotype_chr[, use]
  snpsum.col <- snpsum.col[use, ]
  
  ## hardy-weisburg check
  hardy <- 1e-6
  HWEuse <- with(snpsum.col, (!is.na(z.HWE) & (abs(z.HWE) < abs(qnorm(hardy/2)))))
  HWEuse[is.na(HWEuse)] <- FALSE
  
  cat(ncol(genotype_chr) - sum(HWEuse), "SNPs will be removed due to violation of HWE. \n")
  
  ## subset SNPs that passed the criteria of HWE
  genotype_chr <- genotype_chr[, HWEuse]
  snpsum.col <- snpsum.col[HWEuse, ]
  
  cat(ncol(genotype_chr), " SNPs left after QC. \n")
  genoBIM = genoBIM[colnames(genotype_chr), ]
  
  
  # flip 0 and 2, since snpMatrix coded 0 as double minor allele, 2 as double major allele
  flip.func <- function(snp.id){
    geno.tmp <- genotype_chr[, snp.id]
    snp.summary <- col.summary(geno.tmp)
    if(snp.summary[1, 6] < snp.summary[1, 8]){
      # if 0 and 2 coding is opposite, then switch them back
      geno.tmp <- 2 - as(geno.tmp, "numeric")
    } else {
      geno.tmp <- as(geno.tmp, "numeric")
    }
    return(geno.tmp)
  }
  genotype = flip.func(colnames(genotype_chr))
  
  save(genotype, file = paste0("X_Rdata/chr",chr,"_SNPs.RData"))
  
  print(paste0("finished: chr",chr))
}
print(Sys.time()-start.time)

system("dx upload generate_X_RDATA.R")
system("dx upload -r X_Rdata")

#save(genonum_chr, genoFAM, genoBIM, snpsum.col, file = "C:/Users/wd278/Desktop/HCP_chr22_data.RData")
# the set of SNPs generated by the above code is slightly different then Wei Dai generated back then, which is stored at
# "/gpfs/gibbs/project/zhang_heping/wd278/Project_with_others/Yisha/Data/Genotype/chr1/HCP_afterQC_chr1.RData"
